### PR TITLE
Feature/us4-recipe-details

### LINKED ISSUE #51: US4-FE#2 - Add reviews of other users

### LINKED ISSUE #56: US4-FE#7 - Post a review in the recipe

### LINKED ISSUE #147: US4-BE#3 Create Review models

### LINKED ISSUE #148: US8-FE#1 Add Rating and Review
This should be at the bottom of the recipe details, but right above the list of reviews.
When the review is sent show feedback to the user as confirmation.
Reviews should have a character limit (let's say 2000 for now, ~10 tweets thread lmao)

### LINKED ISSUE #149: US8-FE#2 Add Star Rating
Mimic difficulty, but I would expand it so it can use half stars.

### LINKED ISSUE #150: US8-FE#3 Show review list

### LINKED ISSUE #152: US8-FE#5 Review user feedback
The user should be able to mark a review as helpful or not.

### LINKED ISSUE #159: US4-FE#100 Back button
Since recipe detail can be accessed from multiple places, a proper back button needs to be added.

Every page that can access recipe detail needs to pass its name. Then, the back button just has to navigate to that name.

### PR DESCRIPTION
**Pull request by ivanmansilla**
I have created two new components. Now we have the possibility for users to add reviews of the recipes. At first it was programmed to be in a separate column where they appear directly, but finally they appear in an Offcanvas

### CODE REVIEW

**Comment by ivanmansilla** [comment]
The first version, it was comments, and it looked like this:
image

**Comment by ivanmansilla** [comment]
At this moment, I have modified and now it is an Offcanvas, I have aligned with develop and I have set it to appear on the right when click the button - Toggle Comment:
image

**Comment by ivanmansilla** [comment]
To do so far, what I have done is to create a new component called Comments, which is the reviews of the people, thanks to a map where it is showing them. Right now a mock is used but it is prepared for when we make the fetch bringing the reviews it shows them. 
If we click on the post comment button (it will be review) it shows us a modal, which is another new component that I have created called PostComment.
This is its visualization. The username, will come by props and we will not ask it to the user, and the rating will be added.
image

**Comment by BlessedLongsword** [comment]
I don't know if @xvxnoah already told you this. But in the post comment modal I would make the following changes:

Post a comment -> Make a Review
Username -> Rating (instead of a written field, it should use a star system).

Does the file selection appear in spanish because it uses browser/system language?

**Comment by ivanmansilla** [comment]
Several changes have been made. In short, we already use the post and get for the reviews. Users can now post a review only, with an image or an optional comment, and also rate it from 1 to 5 stars. Also the get brings the information and is responsive. Then we will add a button to like the review for example.
It has also been updated as @BlessedLongsword commented, now it is review and in the post review the username field does not appear.
Comment that the first comments did not appear in the review because they were tests and we did not send them properly.
image

**Comment by ivanmansilla** [comment]
I have implemented the likes and all its functionality. It has cost me a lot, because it has much play with endpoints and various things, but finally I managed to add the functionality of likes, I tried to control all possible errors, by default will be 0 if you do not have likes. A user can only like a review once, and you can't like your own review. Attached is an image of how it looks like and another one of the error that pops up when you try to like something you have already liked.
image
image

**Comment by xvxnoah** [comment]
> I have implemented the likes and all its functionality. It has cost me a lot, because it has much play with endpoints and various things, but finally I managed to add the functionality of likes, I tried to control all possible errors, by default will be 0 if you do not have likes. A user can only like a review once, and you can't like your own review. Attached is an image of how it looks like and another one of the error that pops up when you try to like something you have already liked. image image

Nice work Ivan! I have a couple of comments for you:

1. Can we add some margin between the number of likes and the icon?
2. Can we change the position of the like icon with the number? So we have the icon on the left and the number on its right side. That way we mirror the behaviour of the rating with the stars.
3. Maybe instead of showing that kind of error when trying to like again the review, we can just unlike the review (I will add that endpoint now to the backend as a hotfix).
4. Can we show the errors in a different way? It is just that I don't like to show the errors like web notifications.

**Comment by xvxnoah** [comment]
> > I have implemented the likes and all its functionality. It has cost me a lot, because it has much play with endpoints and various things, but finally I managed to add the functionality of likes, I tried to control all possible errors, by default will be 0 if you do not have likes. A user can only like a review once, and you can't like your own review. Attached is an image of how it looks like and another one of the error that pops up when you try to like something you have already liked. image image
> 
> Nice work Ivan! I have a couple of comments for you:
> 
> 1. Can we add some margin between the number of likes and the icon?
> 2. Can we change the position of the like icon with the number? So we have the icon on the left and the number on its right side. That way we mirror the behaviour of the rating with the stars.
> 3. Maybe instead of showing that kind of error when trying to like again the review, we can just unlike the review (I will add that endpoint now to the backend as a hotfix).
> 4. Can we show the errors in a different way? It is just that I don't like to show the errors like web notifications.

By the way, I think the whole team would love a button to go back to the previous page in recipe details :))

**Comment by ivanmansilla** [comment]
Perfect, I'll start with the comments you have told me! Regarding the issue of dislike I'm waiting and with the theme that comes out something other than an alert, I'll try to investigate something else, to see if I find something cool

**Comment by ivanmansilla** [comment]
I have solved all the comments you made @noah. I have left a margin between icon and number of likes, which is now inverted. It is aligned next to the stars to make it look good. I have also changed the icon to heart, since the dislike is implemented. So when someone likes you, you can click again to remove your like without problems. Also apart from the dislikes I have implemented the arrow which takes you to the userfeed.
I attach images:
Like:
image
Dislike:
image
Arrow:
image

**Comment by xvxnoah** [comment]
> I have solved all the comments you made @noah. I have left a margin between icon and number of likes, which is now inverted. It is aligned next to the stars to make it look good. I have also changed the icon to heart, since the dislike is implemented. So when someone likes you, you can click again to remove your like without problems. Also apart from the dislikes I have implemented the arrow which takes you to the userfeed. I attach images: Like: image Dislike: image Arrow: image

Good job Ivan! I will list here all the "bugs" and corrections that I think will be necessary for this sprint, we can also discuss them tonight after the daily:

1. At the moment a user that is not logged in can like and "unlike" a review.
2. If a user is not logged in, the "Post review" button should not be visible.
3. You should add the average rating of the recipe to the details.
4. The modal for posting a recipe looks good, but it would be nice that followed the same color pattern than the Post Recipe modal in the user feed.
5. As we commented before, you shouldn't be able to like your own review.
6. Users should be able to click on the review's image to see it bigger.
7. You should control how the review text is displayed because as you can see in this screenshot, the username text is on top the review:
Screenshot 2023-11-23 at 18 32 26
8. For the button of Post review I would maybe follow the same colour as the button for posting a recipe.
9. We should consider using a different logo for the stars we want to give in the review, because now is the same logo as the difficulty in the recipe details, which might lead to confusions.

File: package.json [hunk 1]
```diff
@@ -3,6 +3,8 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@fortawesome/free-solid-svg-icons": "^6.4.2",
```
Comment (guzmanalejandro) [inline package.json:6]: We can use more icons of Fortawesome in the next sprint

File: src/components/LikesReview.jsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        const data = await response.json();
+        if (!response.ok) {
+          console.error("Error al enviar el dislike:", data);
+        }
+      } catch (error) {
+        console.error("Error en la solicitud de dislike:", error);
+      }
+    }else{
+      
+      console.log(">>>By2: ", hasLikedByUser)
```
Comment (guzmanalejandro) [inline src/components/LikesReview.jsx]: Are these logs necessary after merging the pull request?

File: src/components/LikesReview.jsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    }
+  }
+    reloadReviews();
+  };
+
+  return (
+    <Row>
+      <Col sm={3}>
+        <FontAwesomeIcon
+          icon={faHeart}
```
Comment (guzmanalejandro) [inline src/components/LikesReview.jsx:74]: In the next sprint we could modify the like to change the empty heart icon to a filled one by liking it, for example.

File: src/components/PostReview.jsx [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        })
+        .catch((error) => console.error("Error:", error));
+  }, [id])
+
+  const handleFileChange = (event) => {
+    const selectedImage= event.target.files[0];
+    setImage(selectedImage);
+  };
+
+  const handlePostReview = async () => {
```
Comment (guzmanalejandro) [inline src/components/PostReview.jsx:42]: In this method we should control if the review is empty (the text) or not

Comment (guzmanalejandro) [reply src/components/PostReview.jsx:42]: Or control it from the backend and launch a popup, but I think it's more feasible to control it in the frontend

File: src/components/LikesReview.jsx (hunk 2 above)
Comment (ivanmansilla) [reply src/components/LikesReview.jsx]: Yes,  I have forgotten to remove it

File: src/components/PostReview.jsx (hunk 4 above)
Comment (ivanmansilla) [reply src/components/PostReview.jsx:42]: The text is optional, can be empty

**Comment by ivanmansilla** [comment]
I fixed all things that PO sayme, here the proves and recipeDetail finally:
image
image
image
not logged:
image

**Review by xvxnoah — APPROVED** [review]
Looks awesome :))))

**Review by xvxnoah — APPROVED** [review]

**Pull request merged**